ra-tls/src/attestation.rs

```diff
@@ -193,39 +202,198 @@ impl Attestation {
+            serde_json::from_slice(&raw_event_log).context("invalid event log")?
+        } else {
+            vec![]
+        };
+        Ok(Self {
+            quote,
+            raw_event_log,
+            event_log,
+            report: (),
+        })
+    }
+
+    /// Extract attestation data from a certificate
+    pub fn from_cert(cert: &impl CertExt) -> Result<Option<Self>> {
+        Self::from_ext_getter(|oid| cert.get_extension_bytes(oid))
+    }
+
+    /// From an extension getter
+    pub fn from_ext_getter(
+        get_ext: impl Fn(&[u64]) -> Result<Option<Vec<u8>>>,
+    ) -> Result<Option<Self>> {
+        let quote = match get_ext(oids::PHALA_RATLS_QUOTE)? {
+            Some(v) => v,
+            None => return Ok(None),
+        };
+        let raw_event_log = get_ext(oids::PHALA_RATLS_EVENT_LOG)?.unwrap_or_default();
+        Self::new(quote, raw_event_log).map(Some)
+    }
+
+    /// Extract attestation from x509 certificate
+    pub fn from_der(cert: &[u8]) -> Result<Option<Self>> {
+        let (_, cert) = parse_x509_certificate(cert).context("Failed to parse certificate")?;
+        Self::from_cert(&cert)
+    }
+
+    /// Extract attestation from x509 certificate in PEM format
+    pub fn from_pem(cert: &[u8]) -> Result<Option<Self>> {
+        let (_, pem) =
+            x509_parser::pem::parse_x509_pem(cert).context("Failed to parse certificate")?;
+        let cert = pem.parse_x509().context("Failed to parse certificate")?;
+        Self::from_cert(&cert)
+    }
+
+    /// Verify the quote
+    pub async fn verify_with_ra_pubkey(
+        self,
+        ra_pubkey_der: &[u8],
+        pccs_url: Option<&str>,
+    ) -> Result<VerifiedAttestation> {
+        self.verify(
+            &QuoteContentType::RaTlsCert.to_report_data(ra_pubkey_der),
+            pccs_url,
+        )
+        .await
+    }
+
+    /// Verify the quote
+    pub async fn verify(
+        self,
+        report_data: &[u8; 64],
+        pccs_url: Option<&str>,
+    ) -> Result<VerifiedAttestation> {
+        let quote = &self.quote;
+        if &self.decode_report_data()? != report_data {
+            anyhow::bail!("report data mismatch");
+        }
+        let report = qvl::collateral::get_collateral_and_verify(quote, pccs_url)
+            .await
+            .context("Failed to get collateral")?;
+        if let Some(report) = report.report.as_td10() {
+            // Replay the event logs
+            let rtmrs = self
+                .replay_event_logs(None)
+                .context("Failed to replay event logs")?;
+            if rtmrs != [report.rt_mr0, report.rt_mr1, report.rt_mr2, report.rt_mr3] {
+                anyhow::bail!("RTMR mismatch");
+            }
         }
-        Ok(())
+        Ok(VerifiedAttestation {
+            quote: self.quote,
+            raw_event_log: self.raw_event_log,
+            event_log: self.event_log,
+            report,
+        })
```

Comment:
Review Hint: This is only place creating `VerifiedAttestation` instance. So when you see `VerifiedAttestation` in the codebase, it must be verified.